service/remotecfg/remotecfg.go

```diff
@@ -0,0 +1,289 @@
+
+	ctrl service.Controller
+
+	mut               sync.RWMutex
+	asClient          agentv1connect.AgentServiceClient
+	ticker            *time.Ticker
+	dataPath          string
+	currentConfigHash string
+}
+
+// ServiceName defines the name used for the remotecfg service.
+const ServiceName = "remotecfg"
+
+// Options are used to configure the remotecfg service. Options are
+// constant for the lifetime of the remotecfg service.
+type Options struct {
+	Logger      log.Logger // Where to send logs.
+	StoragePath string     // Where to cache configuration on-disk.
+}
+
+// Arguments holds runtime settings for the remotecfg service.
+type Arguments struct {
+	URL              string                   `river:"url,attr,optional"`
+	ID               string                   `river:"id,attr,optional"`
+	Metadata         map[string]string        `river:"metadata,attr,optional"`
+	PollFrequency    time.Duration            `river:"poll_frequency,attr,optional"`
+	HTTPClientConfig *config.HTTPClientConfig `river:",squash"`
+}
+
+// GetDefaultArguments populates the default values for the Arguments struct.
+func GetDefaultArguments() Arguments {
+	return Arguments{
+		ID:               agentseed.Get().UID,
+		Metadata:         make(map[string]string),
+		PollFrequency:    1 * time.Minute,
+		HTTPClientConfig: config.CloneDefaultHTTPClientConfig(),
+	}
+}
+
+// SetToDefault implements river.Defaulter.
+func (a *Arguments) SetToDefault() {
+	*a = GetDefaultArguments()
+}
+
+// Validate implements river.Validator.
+func (a *Arguments) Validate() error {
+	// We must explicitly Validate because HTTPClientConfig is squashed and it
+	// won't run otherwise
+	if a.HTTPClientConfig != nil {
+		return a.HTTPClientConfig.Validate()
+	}
+
+	return nil
+}
+
+// New returns a new instance of the remotecfg service.
+func New(opts Options) (*Service, error) {
+	basePath := filepath.Join(opts.StoragePath, ServiceName)
+	err := os.MkdirAll(basePath, 0750)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Service{
+		opts:     opts,
+		asClient: noopClient{},
+		ticker:   time.NewTicker(math.MaxInt64),
+	}, nil
+}
+
+// Data does not expose anything for the remotecfg service during runtime.
+func (s *Service) Data() any {
+	return nil
+}
+
+// Definition returns the definition of the remotecfg service.
+func (s *Service) Definition() service.Definition {
+	return service.Definition{
+		Name:       ServiceName,
+		ConfigType: Arguments{},
+		DependsOn:  nil, // remotecfg has no dependencies.
+	}
+}
+
+var _ service.Service = (*Service)(nil)
+
+// Run implements [service.Service] and starts the remotecfg service. It will
+// run until the provided context is canceled or there is a fatal error.
+func (s *Service) Run(ctx context.Context, host service.Host) error {
+	s.ctrl = host.NewController(ServiceName)
+
+	// Let's try to read from the API or the local cache as a fallback.
+	var b1, b2 []byte
+	b1, err := s.getAPIConfig()
+	b2, err2 := s.getCachedConfig()
```

Comment:
I've been looking at this PR again after reading @wildum's [comment](https://github.com/grafana/agent/pull/6277#discussion_r1472607048), trying to make sense of when cache invalidation should occur. 

I came up with two cases that I think are the most important:

* The existing cache should be invalidated when the `remotecfg` block is no longer being configured. 
* The existing cache should be invalidated when the `remotecfg` block is changed to point to a different server, or with a different authentication username, or potentially if the agent's identifying labels have changed. 

I don't think I have a fully exhaustive list for the second case, but it seems like the items I listed all fall into the category of "the endpoint is different or the agent's identity is different."